imitation/model_env.py

```diff
@@ -0,0 +1,345 @@
+"""Finite-horizon discrete environments with known transition dynamics. These
+are handy when you want to perform exact maxent policy optimisation."""
+
+import abc
+
+import gym
+from gym import spaces
+import numpy as np
+
+
+class ModelBasedEnv(gym.Env, abc.ABC):
+    """ABC for tabular environments with known dynamics."""
+    def __init__(self):
+        self.cur_state = None
+        self.n_actions_taken = None
+        # we must cache action & observation spaces instead of reconstructing
+        # anew so that the random state of the action space (!!) is preserved
+        self._action_space = None
+        self._observation_space = None
+        self.seed()
+
+    @property
+    def action_space(self):
+        if not self._action_space:
+            self._action_space = spaces.Discrete(self.n_actions)
+        return self._action_space
+
+    @property
+    def observation_space(self):
+        if not self._observation_space:
+            self._observation_space = spaces.Box(low=float('-inf'),
+                                                 high=float('inf'),
+                                                 shape=(self.obs_dim, ))
+        return self._observation_space
+
+    def seed(self, seed=None):
+        if seed is None:
+            seed = np.random.randint(0, 1 << 31)
+        self.rand_state = np.random.RandomState(seed=seed)
+        return [seed]
+
+    def reset(self):
+        self.cur_state = 0
+        self.n_actions_taken = 0
+        # as in step(), we copy so that it can't be mutated in-place (updates
+        # will be reflected in self.observation_matrix!)
+        return self.observation_matrix[self.cur_state].copy()
+
+    def step(self, action):
+        assert self.cur_state is not None \
+            and self.n_actions_taken is not None, \
+            "remember to call reset() before first step()"
+        old_state = self.cur_state
+        out_dist = self.transition_matrix[old_state, action]
+        choice_states = np.arange(self.n_states)
+        next_state = int(
+            self.rand_state.choice(choice_states, p=out_dist, size=()))
+        self.cur_state = next_state
+        self.n_actions_taken += 1
+        done = self.n_actions_taken >= self.horizon
+        reward = self.reward_matrix[old_state]
```

Comment:
Given the reward is currently state-only, this does feel odd to me. However, it would make complete sense for a state-action reward, since we do not know the action until this point. So I think it's probably best to keep it as-is so we have an upgrade path to state-action or state-action-successor state rewards in the future.